initial contribution of IPP binding

peuter opened the pull request:
including mDNS-based disovery service

kaikreuzer commented:
Thanks - I'll try to review it soon!

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+        xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+    <name>Cups Binding</name>
+    <description>This is the binding for Cups.</description>
```
Better briefly explain what Cups is about, i.e. "for the Common Unix Printing Service (CUPS)"

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+		<config-description>
+			<parameter name="host" type="text">
+				<label>Hostname</label>
+				<description>Hostname or IP address of the pulseaudio server</description>
```
I see which binding you used as a template ;-)
Please fix c&p error

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+			</parameter>
+			<parameter name="port" type="integer">
+				<label>Port</label>
+				<description>Port of the pulseaudio server</description>
```
dito

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+				<label>Hostname</label>
+				<description>Hostname or IP address of the pulseaudio server</description>
+				<default>localhost</default>
+				<required>true</required>
```
"required" should be put as an attribute, not an element anymore.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+				<label>Port</label>
+				<description>Port of the pulseaudio server</description>
+				<default>4712</default>
+				<required>false</required>
```
"required" should be put as an attribute, not an element anymore.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+			<parameter name="port" type="integer">
+				<label>Port</label>
+				<description>Port of the pulseaudio server</description>
+				<default>4712</default>
```
Is this really the default here or also left over from pulseaudio?

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+			</parameter>
+			<parameter name="refreshInterval" type="integer">
+				<label>Refresh Interval</label>
+				<description>The refresh interval in ms which is used to poll given pulseaudio server.</description>
```
c&p error

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="bridge" />
+		</supported-bridge-type-refs>
+		<label>Cups Printer Thing</label>
```
Just "Printer" should be fine.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+				<label>Refresh Interval</label>
+				<description>The refresh interval in ms which is used to poll given pulseaudio server.</description>
+				<default>30000</default>
+				<required>false</required>
```
"required" should be put as an attribute, not an element anymore.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+			<parameter name="name" type="text">
+				<label>Name</label>
+				<description>The name of the printer</description>
+				<required>true</required>
```
"required" should be put as an attribute, not an element anymore.

kaikreuzer replied:
Is the name really mandatory and not the url maybe enough (from which a name could be derived)?

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+			<parameter name="url" type="text">
+				<label>URL</label>
+				<description>The URL of the printer</description>
+				<required>true</required>
```
"required" should be put as an attribute, not an element anymore.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+	</channel-type>
+	<channel-type id="waitingJobs">
+		<item-type>Number</item-type>
+		<label>Jobs</label>
```
Should have a different label than the "jobs" channel.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,76 @@
+	</channel-type>
+	<channel-type id="doneJobs">
+		<item-type>Number</item-type>
+		<label>Jobs</label>
```
Should have a different label than the "jobs" channel.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,38 @@
+ org.openhab.core.library.types,
+ org.osgi.framework;version="1.8.0",
+ org.slf4j
+Service-Component: OSGI-INF/*
```
Please change to OSGI-INF/*.xml

kaikreuzer commented on addons/binding/org.openhab.binding.cups/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,38 @@
+ org.eclipse.smarthome.core.types,
+ org.eclipse.smarthome.io.transport.mdns.discovery,
+ org.openhab.core.library.types,
+ org.osgi.framework;version="1.8.0",
```
please remove version constraint

kaikreuzer commented on addons/binding/org.openhab.binding.cups/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,38 @@
+Import-Package: com.google.common.collect,
+ javax.jmdns,
+ org.apache.commons.codec,
+ org.apache.commons.lang;version="2.6.0",
```
please remove version constraint

kaikreuzer commented on addons/binding/org.openhab.binding.cups/pom.xml:
```diff
@@ -0,0 +1,24 @@
+
+  <parent>
+    <groupId>org.openhab</groupId>
+    <artifactId>binding</artifactId>
```
this looks like an outdated parent - it should be groupId=org.openhab.binding and artifactId=pom

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CupsBridgeHandler extends BaseBridgeHandler {
```
Please add a short JavaDoc here, including an author tag.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+public class CupsBridgeHandler extends BaseBridgeHandler {
+	private Logger logger = LoggerFactory.getLogger(CupsBridgeHandler.class);
+
+	public String host;
```
should these three really be public?

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+				client = new CupsClient(host, port);
+				startAutomaticRefresh();
+			} catch (Exception e) {
+				logger.error("Couldn't connect to Cups server [Host '"
```
It cannot connect at the moment. But maybe the server is down just now and back in a minute? So probably you should also schedule a job that regularly retries if it can establish a server communication.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+
+	@Override
+	public void dispose() {
+		client = null;
```
You should also cancel any scheduled jobs in here.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+	public boolean registerDeviceStatusListener(
+			PrinterStatusListener deviceStatusListener) {
+		if (deviceStatusListener == null) {
+			throw new NullPointerException(
```
This should rather be an IllegalArgumentException

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+		}
+		boolean result = deviceStatusListeners.add(deviceStatusListener);
+		if (result) {
+			// onUpdate();
```
why is this commented out? Either activate or remove it please

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsBridgeHandler.java:
```diff
@@ -0,0 +1,171 @@
+			PrinterStatusListener deviceStatusListener) {
+		boolean result = deviceStatusListeners.remove(deviceStatusListener);
+		if (result) {
+			// onUpdate();
```
dito

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+ * 
+ * @author Tobias Braeutigam - Initial contribution
+ */
+public class CupsHandler extends BaseThingHandler implements
```
probably better rename to CupsPrinterHandler?

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+			refreshJob.cancel(true);
+			refreshJob = null;
+		}
+		updateStatus(ThingStatus.OFFLINE);
```
you do not need (or rather must not) update the status here anymore - the framework takes care of this.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+			refreshJob = null;
+		}
+		updateStatus(ThingStatus.OFFLINE);
+		logger.debug("Thing {} {} disposed.", getThing().getUID(), url);
```
it's not the thing, but the handler that gets disposed here!

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+	public void handleCommand(ChannelUID channelUID, Command command) {
+		if (command instanceof RefreshType) {
+			CupsBridgeHandler bridge = getBridgeHandler();
+			bridge.handleCommand(channelUID, command);
```
you might want to do a null check here

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+			try {
+				updateState(new ChannelUID(getThing().getUID(), CupsBindingConstants.JOBS_CHANNEL), new DecimalType(device.getJobs(WhichJobsEnum.ALL, "", false).size()));
+			} catch (Exception e) {
+				logger.error("error updating jobs channel, reason: {}",e.getMessage());
```
reduce to debug level

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+			try {
+				updateState(new ChannelUID(getThing().getUID(), CupsBindingConstants.WAITING_JOBS_CHANNEL), new DecimalType(device.getJobs(WhichJobsEnum.NOT_COMPLETED, "", false).size()));
+			} catch (Exception e) {
+				logger.error("error updating waiting-jobs channel, reason: {}",e.getMessage());
```
reduce to debug level

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+			try {
+				updateState(new ChannelUID(getThing().getUID(), CupsBindingConstants.DONE_JOBS_CHANNEL), new DecimalType(device.getJobs(WhichJobsEnum.COMPLETED, "", false).size()));
+			} catch (Exception e) {
+				logger.error("error updating done-jobs channel, reason: {}",e.getMessage());
```
reduce to debug level

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/handler/CupsHandler.java:
```diff
@@ -0,0 +1,176 @@
+	@Override
+	public void onDeviceRemoved(Bridge bridge, CupsPrinter device) {
+		if (device.getPrinterURL().equals(url)) {
+			bridgeHandler.unregisterDeviceStatusListener(this);
```
can you be sure that the bridgeHandler has always been set here?

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/CupsHandlerFactory.java:
```diff
@@ -0,0 +1,114 @@
+	@Override
+	public Thing createThing(ThingTypeUID thingTypeUID,
+			Configuration configuration, ThingUID thingUID, ThingUID bridgeUID) {
+		logger.info("createThing(" + thingTypeUID + "," + configuration + ","
```
please reduce to debug level

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/CupsHandlerFactory.java:
```diff
@@ -0,0 +1,114 @@
+				+ thingUID + "," + bridgeUID + ")");
+		if (CupsBindingConstants.SUPPORTED_BRIDGE_TYPES_UIDS
+				.contains(thingTypeUID)) {
+			logger.info("creating bridge {}", thingTypeUID);
```
please reduce to debug level

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/CupsHandlerFactory.java:
```diff
@@ -0,0 +1,114 @@
+		if (CupsBindingConstants.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+			ThingUID deviceUID = getCupsPrinterUID(thingTypeUID, thingUID,
+					configuration, bridgeUID);
+			logger.info("creating thing " + thingTypeUID
```
please reduce to debug level

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/CupsHandlerFactory.java:
```diff
@@ -0,0 +1,114 @@
+	private void registerDeviceDiscoveryService(CupsBridgeHandler bridge) {
+		CupsDeviceDiscoveryService discoveryService = new CupsDeviceDiscoveryService();
+		discoveryService.activate(bridge);
+		this.discoveryServiceReg = bundleContext.registerService(
```
There should be one such discoveryServiceReg per bridge and not just one in total, right? So better keep them in a Map instead.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/CupsHandlerFactory.java:
```diff
@@ -0,0 +1,114 @@
+
+	@Override
+	protected void removeHandler(ThingHandler thingHandler) {
+		if (this.discoveryServiceReg != null) {
```
only unregister service discovery for the given bridge handler.

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/discovery/PrinterDiscoveryParticipant.java:
```diff
@@ -0,0 +1,99 @@
+ * @author Tobias Bräutigam
+ *
+ */
+public class PrinterDiscoveryParticipant implements MDNSDiscoveryParticipant {
```
Shouldn't this be rather a CupsServerDiscoveryParticipant instead of a PrinterDiscoveryParticipant?

kaikreuzer commented on addons/binding/org.openhab.binding.cups/src/main/java/org/openhab/binding/cups/internal/discovery/PrinterDiscoveryParticipant.java:
```diff
@@ -0,0 +1,99 @@
+			try {
+				logger.trace("testing connection to cups server {}:{}",
+						hostname,port);
+				CupsClient client = new CupsClient(hostname,port);
```
Discovery should be as passive as possible, so I do not like creating a client connection here. What does it bring after all? Imho, if a CUPS server announces itself through mDNS, this is good enough to list it as an Inbox entry. It should also do this if there are no printers available on the server - maybe they are added later only.

kaikreuzer commented:
Thanks @peuter, I have finished my review.

As the binding contains an external lib that is not under EPL, but LGPL (as far as I can tell), I'd like to ask you to also put an "about.html" file inside the bundle that provides details about the licensing. Please find a how-to and a template for such a file at https://www.eclipse.org/legal/epl/about.php.

One general remark: You are mostly using string concatenation in your logging statements. Performance wise it would be better to go for parameterized logging, see explanation at http://www.slf4j.org/faq.html#logging_performance

Please leave me a comment as soon as you have addressed my comments - thanks!

peuter commented:
I´m sorry but i totally messed up this branch now, by rebasing/merging/squashing. Not only that all my changes due to your code review are gone, but I lost the complete cups-binding source code and have no Idea how to bring it back (I also have no idea how it got lost either)

kaikreuzer commented:
Oh oh...
I have just pushed your branch that I reviewed yesterday to my fork: https://github.com/kaikreuzer/openhab2/tree/peuter-cups-binding
You should be able to take it back from there (but of course it does not include your latest changes after my review)

peuter commented:
Thank you, that woudl help but I don´t see the cups-binding in this branch:
https://github.com/kaikreuzer/openhab2/tree/peuter-cups-binding/addons/binding

kaikreuzer commented:
Sorry - now you should see it!

peuter commented:
Ok got it, thanks a lot.

peuter commented:
If you dont mind I would close this PR and start with a new branch + PR for the binding. I don´t see any chance to clean up this messy branch.

peuter commented:
Ok I think I got it (even in this branch) ;-)

kaikreuzer commented:
Thanks, seems as if you have made it :-)
I was just trying to test it actively and I came across a couple of exceptions in my log (while in general the binding seems to work). Do you have any feedback on these?

Here seems to be a null check missing:
```
22:23:44.297 [DEBUG] [b.c.handler.CupsPrinterHandler:100  ] - Exception occurred during execution: nulljava.lang.NullPointerException: null
	at org.openhab.binding.cups.handler.CupsPrinterHandler$1.run(CupsPrinterHandler.java:94)
```

This one is probably deeper down in the cups library. Maybe the http client needs to be parameterized in some other way to avoid this problem? ("Kona" is the hostname of my cups server, which had been automatically discovered by the cups binding)
```
22:23:37.556 [DEBUG] [b.c.handler.CupsPrinterHandler:100  ] - Exception occurred during execution: Kona: nodename nor servname provided, or not knownjava.net.UnknownHostException: Kona: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:901)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1293)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1246)
	at java.net.InetAddress.getAllByName(InetAddress.java:1162)
	at java.net.InetAddress.getAllByName(InetAddress.java:1098)
	at org.apache.http.impl.conn.SystemDefaultDnsResolver.resolve(SystemDefaultDnsResolver.java:45)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.resolveHostname(DefaultClientConnectionOperator.java:278)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:162)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:294)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:643)
```
Probably related, but yet another logging entry:
```
22:20:21.280 [ERROR] [cups.handler.CupsBridgeHandler:98   ] - exception while getting printer from org.cups4j.CupsClient@6b6c313: Kona
```
No exception, but a bit weird that it is discovered 3 times:
```
22:20:44.373 [DEBUG] [CupsServerDiscoveryParticipant:52   ] - Discovered a cups server bridge with name 'Kona'
22:20:44.374 [DEBUG] [CupsServerDiscoveryParticipant:88   ] - Created a DiscoveryResult for cups server bridge on host 'Kona'
22:20:44.374 [DEBUG] [CupsServerDiscoveryParticipant:52   ] - Discovered a cups server bridge with name 'Kona'
22:20:44.375 [DEBUG] [CupsServerDiscoveryParticipant:88   ] - Created a DiscoveryResult for cups server bridge on host 'Kona'
22:20:44.375 [DEBUG] [CupsServerDiscoveryParticipant:52   ] - Discovered a cups server bridge with name 'Kona'
22:20:44.376 [DEBUG] [CupsServerDiscoveryParticipant:88   ] - Created a DiscoveryResult for cups server bridge on host 'Kona'
22:20:44.377 [DEBUG] [s.c.d.internal.PersistentInbox:134  ] - Discovery result with thing 'cups:bridge:Kona' not added as inbox entry. It is already present as thing in the ThingRegistry.
22:20:44.377 [DEBUG] [s.c.d.internal.PersistentInbox:140  ] - The configuration for thing 'cups:bridge:Kona' is updated...
```

peuter commented:
Hm, I don´t see any of these errors. Anyway I found a typo in the 
thing-types.xml and added a null check to the CupsPrinterHandler.
The errors you found seem to be a DNS problem, can you ping the Kona 
server from you openHAB test server?

I just found another problem. I have a real network printer which also 
listens to port 631 (and annouces that via mDNS), so it is added as a 
cups-bridge, which it is not (that why I added the connection check in 
the discovery, to check if it is a real cups server).
So IMHO there are 2 solutions for this problem:
1. Check if the Server is connectable before adding the bridge as a 
DiscoveryResult (like it was implemented before)
2. As the cups-bridges are of no use to the user: Is there anyway to 
hide them from the list of DiscoveryResults and just use them in the 
background to discover the printers.

Am 12.04.2015 um 22:31 schrieb Kai Kreuzer:
>
> Thanks, seems as if you have made it :-)
> I was just trying to test it actively and I came across a couple of 
> exceptions in my log (while in general the binding seems to work). Do 
> you have any feedback on these?
>
> Here seems to be a null check missing:
>
> |22:23:44.297 [DEBUG] [b.c.handler.CupsPrinterHandler:100  ] - Exception occurred during execution: nulljava.lang.NullPointerException: null
>      at org.openhab.binding.cups.handler.CupsPrinterHandler$1.run(CupsPrinterHandler.java:94)
> |
>
> This one is probably deeper down in the cups library. Maybe the http 
> client needs to be parameterized in some other way to avoid this 
> problem? ("Kona" is the hostname of my cups server, which had been 
> automatically discovered by the cups binding)
>
> |22:23:37.556 [DEBUG] [b.c.handler.CupsPrinterHandler:100  ] - Exception occurred during execution: Kona: nodename nor servname provided, or not knownjava.net.UnknownHostException: Kona: nodename nor servname provided, or not known
>      at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
>      at java.net.InetAddress$1.lookupAllHostAddr(InetAddress.java:901)
>      at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1293)
>      at java.net.InetAddress.getAllByName0(InetAddress.java:1246)
>      at java.net.InetAddress.getAllByName(InetAddress.java:1162)
>      at java.net.InetAddress.getAllByName(InetAddress.java:1098)
>      at org.apache.http.impl.conn.SystemDefaultDnsResolver.resolve(SystemDefaultDnsResolver.java:45)
>      at org.apache.http.impl.conn.DefaultClientConnectionOperator.resolveHostname(DefaultClientConnectionOperator.java:278)
>      at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:162)
>      at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:294)
>      at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:643)
> |
>
> Probably related, but yet another logging entry:
>
> |22:20:21.280 [ERROR] [cups.handler.CupsBridgeHandler:98   ] - exception while getting printer from org.cups4j.CupsClient@6b6c313: Kona
> |
>
> No exception, but a bit weird that it is discovered 3 times:
>
> |22:20:44.373 [DEBUG] [CupsServerDiscoveryParticipant:52   ] - Discovered a cups server bridge with name 'Kona'
> 22:20:44.374 [DEBUG] [CupsServerDiscoveryParticipant:88   ] - Created a DiscoveryResult for cups server bridge on host 'Kona'
> 22:20:44.374 [DEBUG] [CupsServerDiscoveryParticipant:52   ] - Discovered a cups server bridge with name 'Kona'
> 22:20:44.375 [DEBUG] [CupsServerDiscoveryParticipant:88   ] - Created a DiscoveryResult for cups server bridge on host 'Kona'
> 22:20:44.375 [DEBUG] [CupsServerDiscoveryParticipant:52   ] - Discovered a cups server bridge with name 'Kona'
> 22:20:44.376 [DEBUG] [CupsServerDiscoveryParticipant:88   ] - Created a DiscoveryResult for cups server bridge on host 'Kona'
> 22:20:44.377 [DEBUG] [s.c.d.internal.PersistentInbox:134  ] - Discovery result with thing 'cups:bridge:Kona' not added as inbox entry. It is already present as thing in the ThingRegistry.
> 22:20:44.377 [DEBUG] [s.c.d.internal.PersistentInbox:140  ] - The configuration for thing 'cups:bridge:Kona' is updated...
> |
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openhab/openhab2/pull/164#issuecomment-92121355>.
>

kaikreuzer commented:
> Hm, I don´t see any of these errors.

This seems to be a Mac OS X problem: http://crunchify.com/getting-java-net-unknownhostexception-nodename-nor-servname-provided-or-not-known-error-on-mac-os-x-update-your-privateetchosts-file/

> I just found another problem. I have a real network printer which also
listens to port 631 (and annouces that via mDNS)

Doesn't the mDNS announcement contain any details that it is about a cups service? Ah, I just read up on this: The annoucement is actually for IPP (http://en.wikipedia.org/wiki/Internet_Printing_Protocol) and so your printer seems to support this as well. So it means that the cups binding is specific to cups and cannot handle IPP itself? Wouldn't it actually make much sense to have support for IPP in general (as CUPS uses this standard as well)?

peuter commented:
Sure it would make sense to support IPP in general, but there is no 
usable java-lib to do so. CUPS4j is using a subset of IPP and I did not 
find a better IPP lib.
At least the querying for the number of jobs always returns 0 on my 
IPP-Printer. I don´t know, wether this is not supported by IPP, my 
printer or CUPS4j.
So I think we have to stick to CUPS.

Am 13.04.2015 um 12:20 schrieb Kai Kreuzer:
>
>     Hm, I don´t see any of these errors.
>
> This seems to be a Mac OS X problem: 
> http://crunchify.com/getting-java-net-unknownhostexception-nodename-nor-servname-provided-or-not-known-error-on-mac-os-x-update-your-privateetchosts-file/
>
>     I just found another problem. I have a real network printer which also
>     listens to port 631 (and annouces that via mDNS)
>
> Doesn't the mDNS announcement contain any details that it is about a 
> cups service? Ah, I just read up on this: The annoucement is actually 
> for IPP (http://en.wikipedia.org/wiki/Internet_Printing_Protocol) and 
> so your printer seems to support this as well. So it means that the 
> cups binding is specific to cups and cannot handle IPP itself? 
> Wouldn't it actually make much sense to have support for IPP in 
> general (as CUPS uses this standard as well)?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openhab/openhab2/pull/164#issuecomment-92301542>.
>

kaikreuzer commented:
This class (https://svn.apache.org/repos/asf/harmony/enhanced/archive/classlib/java6/modules/print/src/main/java/common/org/apache/harmony/x/print/cups/CUPSClient.java) claims that CUPS is an extended version of IPP - I would therefore expect that cups implicitly also supports IPP-compatible printers, maybe one just have to reduce the available functionality.
Did you check the Apache Harmony client whether this could be an alternative? But maybe cups4j is just fine and it is indeed your printer that is buggy...

peuter commented:
I took my information from this page:
http://www.cups4j.org/index.php?option=com_content&view=article&id=4&Itemid=2

"Cups4J has a focus on <http://www.cups.org/>CUPS <http://www.cups.org/> 
which means it will also support extensions of IPP provided by 
<http://www.cups.org/>CUPS <http://www.cups.org/> . As a consequence it 
is likely that this library will not work perfectly with other IPP 
implementations. At least some features might not be supported."

But i did some further investigation, with my IPP Printer and I think 
that the printer is the problem, because the communication looks good.
I tried to refactor the binding for general IPP support and I´m stuck 
right now, as the printers are discovered but the corresponding things 
are not created.
As the Printers are announced by mDNS, there is no need for a bridge 
(actually is was never needed, it was just easier coding with the bridge).
It would be great of you could find the time to take a look into the 
code and give my a hint what I am missing. I tried to copy from the 
Sonos binding as it does discovery without a bridge, but still did not 
manage to get it working.
https://github.com/peuter/openhab2/tree/ipp-binding/addons/binding/org.openhab.binding.cups

Thanks.

Am 13.04.2015 um 16:25 schrieb Kai Kreuzer:
>
> This class 
> (https://svn.apache.org/repos/asf/harmony/enhanced/archive/classlib/java6/modules/print/src/main/java/common/org/apache/harmony/x/print/cups/CUPSClient.java) 
> claims that CUPS is an extended version of IPP - I would therefore 
> expect that cups implicitly also supports IPP-compatible printers, 
> maybe one just have to reduce the available functionality.
> Did you check the Apache Harmony client whether this could be an 
> alternative? But maybe cups4j is just fine and it is indeed your 
> printer that is buggy...
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openhab/openhab2/pull/164#issuecomment-92377359>.
>

kaikreuzer commented:
It seems I waited too long, the url to your binding does not work anymore...
So in general, having a cups binding instead of an ipp binding makes a lot of sense, because cups is a superset of ipp. So using cups4j should be just fine.
I also agree that IPP printers should be directly added as a thing without any bridge. Bridges then only make sense for cups servers, which provide access to multiple printers at the same time.
Send me the link to your discovery code and I will try to find out why it does not work for your printer.

peuter commented:
Right, I move the binding to:
https://github.com/peuter/openhab2/tree/ipp-binding/addons/binding/org.openhab.binding.ipp

The Discovery part is here:
https://github.com/peuter/openhab2/blob/ipp-binding/addons/binding/org.openhab.binding.ipp/src/main/java/org/openhab/binding/ipp/internal/discovery/IppPrinterDiscoveryParticipant.java

But thats not the problem, discovery is working, the 
createResult()-method is called and the DiscoveryResult gets created, 
but the Thing does not get created in:
https://github.com/peuter/openhab2/blob/ipp-binding/addons/binding/org.openhab.binding.ipp/src/main/java/org/openhab/binding/ipp/internal/IppHandlerFactory.java

Hope you find the problem, I´m pretty sure that it´s some tiny, stupid 
mistake, which I didn´t find yet ;-)


I guess you ment: "..having a ipp binding instead of a cups binding 
makes a lot of sense..."

The bridges make no sense for cups servers either, because all 
communication is done directly with the printers (of course internally 
this is done via the cups server)
 From the CUPS4j point-of-view you use CupsPrinter as a Thing to 
communicate. CupsClient would be the Bridge, but you can only call 
getPrinters() to get a list of CupsPrinters, and that job is already 
done by mDNS, as the Printers are announced there directly. So no need 
to use a Bridge in this context.

Am 15.04.2015 um 10:29 schrieb Kai Kreuzer:
>
> It seems I waited too long, the url to your binding does not work 
> anymore...
> So in general, having a cups binding instead of an ipp binding makes a 
> lot of sense, because cups is a superset of ipp. So using cups4j 
> should be just fine.
> I also agree that IPP printers should be directly added as a thing 
> without any bridge. Bridges then only make sense for cups servers, 
> which provide access to multiple printers at the same time.
> Send me the link to your discovery code and I will try to find out why 
> it does not work for your printer.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openhab/openhab2/pull/164#issuecomment-93259848>.
>

kaikreuzer commented:
> I guess you ment: "..having a ipp binding instead of a cups binding
makes a lot of sense..."

No, I meant what I said. You said "it will also support extensions of IPP", i.e. CUPS can do MORE than IPP and thus your binding can do more. It means that if you support CUPS, you automatilcally support IPP. That's why I thought it should stay the CUPS binding and not be refactored to an IPP binding. This is, unless you say that the extensions of CUPS over IPP are irrelevant and you anyhow do not support them (what are those extensions btw?).

> that job is already
done by mDNS, as the Printers are announced there directly

Does this mean that if a CUPS server has 5 attached printers, you get 5 different mDNS announcements and thus there will be automatically 5 different discovery results that you can approve independent from each other? Yes, if this is the case and there is no such scenario where you might have to authenticate against the CUPS server, then we can indeed do without a Bridge.

> Hope you find the problem, I´m pretty sure that it´s some tiny, stupid
mistake, which I didn´t find yet ;-)

You know that you have to approve the discovery result before a thing is created? Otherwise you indeed have to wait forever. At the moment that you approve the discovery result (e.g. through the console "smarthome inbox approve <id>" or through the PaperUI), your createThing() method should get called.

peuter commented:
> No, I meant what I said. You said "it will also support extensions of 
> IPP", i.e. CUPS can do MORE than IPP and thus your binding can do 
> more. It means that if you support CUPS, you automatilcally support 
> IPP. That's why I thought it should stay the CUPS binding and not be 
> refactored to an IPP binding. This is, unless you say that the 
> extensions of CUPS over IPP are irrelevant and you anyhow do not 
> support them (what are those extensions btw?).
>
Maybe CUPS can do more, but I don´t use that in the binding (I don´t 
even know, in which way CUPS is extending IPP). The binding only reads 
the number of waiting/done print jobs from the printer, and from my 
understanding this is part of IPP.
The IPP naming of the binding should be ok.
>
> Does this mean that if a CUPS server has 5 attached printers, you get 
> 5 different mDNS announcements and thus there will be automatically 5 
> different discovery results that you can approve independent from each 
> other? Yes, if this is the case and there is no such scenario where 
> you might have to authenticate against the CUPS server, then we can 
> indeed do without a Bridge.
>
Yes 5 printers => 5 mDNS announcements and no authentication against server.
>
> You know that you have to approve the discovery result before a thing 
> is created? Otherwise you indeed have to wait forever. At the moment 
> that you approve the discovery result (e.g. through the console 
> "smarthome inbox approve " or through the PaperUI), your createThing() 
> method should get called.
>
The inbox is empty, so there is nothing to approve.

kaikreuzer commented:
> The IPP naming of the binding should be ok.

Ok, fine with that.

> The inbox is empty, so there is nothing to approve.

Set the line https://github.com/openhab/openhab2/blob/master/distribution/openhabhome/conf/services/runtime.cfg#L1 to false, then discovery results should stay in the inbox until they are manually approved.
Note that you need to delete the thing first ("smarthome things clear"), if it now already exists, before you will see new inbox entries for it at all.

peuter commented:
Ok, I think I found a bug in the mDNS-DiscoveryService. The 
thingDiscovered() is only called in serviceAdded()
https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.transport.mdns/src/main/java/org/eclipse/smarthome/io/transport/mdns/discovery/MDNSDiscoveryService.java#L111

but not in initializeParticipants() or addMdnsDiscoveryParticipant(). 
And therefore nothing is discovered here. Shame on me, as I have coded 
this ;-)

peuter commented:
As the binding changed it´s name I suggest I close this PR and create a new one for the IPP-Binding. 
@kaikreuzer  Are you ok with that?

kaikreuzer commented:
However you like - we can also simply rename this PR, then we have the history in one place.

peuter commented:
Ok, I rebased this branch on my ipp-developemt branch, should be ready to merge now.

kaikreuzer commented:
Thanks! I have just released a new stable of ESH, so that the mDNS fix is now available in openHAB 2 and I was able to test this binding.

One observation: From the 3 channels, only "waiting Jobs" seems to be of most interest. For the other two, I don't see too many use cases. So would it maybe make sense to mark the other two channels as "advanced", so that they do not appear in the UI by default? And I would possibly rename "Jobs" in "total jobs".

Another observation: I have a Canon printer, which is discovered but cannot be accessed. This is because hostname is null as the ServiceInfo does not return any value on getServer(). Nonetheless, the ServiceInfo contains the IP address in the field ipv4Addresses. Should maybe the IP addresses by taken instead of the hostname?

peuter commented:
I agree with setting the channels to advanced as they aren´t really of much interest.
I tried to replace the whole usage of hostname in the creation of the discovery result. 
I also changed the ThingUID to use service.getName(), as this should be unique.
What happens if the printer´s ip address changes? The ThingUID does not change so the URL-property gets updated, right?

buildhive commented:
[openhab » openhab2 #4](https://buildhive.cloudbees.com/job/openhab/job/openhab2/4/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
> What happens if the printer´s ip address changes? The ThingUID does not change so the URL-property gets updated, right?

Yes, correct. If you create a new discovery result for the same UID, the properties of the existing thing will be automatically updated by the framework.

kaikreuzer commented:
@peuter Please let me know, if you see a chance to address the problem or if there is any need to discuss on the solution!

peuter commented:
I thought that I solved the problem with the last commit. Have you tested the last version with your printer?

kaikreuzer commented:
No, as you had a question back, I thought you were not yet done. Will test it right away!

kaikreuzer commented:
Unfortunately, this Canon printer does not turn up during mDNS discovery at all anymore... Will have to investigate, why this is the case...
But I have an issue with another printer (which is shared through a Mac):
```
21:34:57.100 [DEBUG] [.ipp.handler.IppPrinterHandler:139  ] - error updating jobs channel, reason: Provider com.sun.xml.internal.bind.v2.ContextFactory could not be instantiated: javax.xml.bind.JAXBException: "ch.ethz.vppserver.schema.ippclient" doesnt contain ObjectFactory.class or jaxb.index

```
Any idea what this is about?

kaikreuzer commented on addons/binding/org.openhab.binding.ipp/src/main/java/org/openhab/binding/ipp/internal/discovery/IppPrinterDiscoveryParticipant.java:
```diff
@@ -0,0 +1,126 @@
+	
+	private String getPrinterName(ServiceInfo service) {
+		String rp = service.getPropertyString("rp");
+		if (rp==null) {
```
Is "rp" really a good default? In my case, "rp" is filled ("printers/HP_LaserJet_M2727nf_MFP__745B02_"), but the service name ("HP LaserJet M2727nf MFP (745B02) @ Kona") would actually be a much better label text. Did you come across situations, where there is a name given, but which is not usable as a label?

peuter replied:
I guess you are right. This is just a "leftover" from the early stages of developing this binding.

buildhive commented:
[openhab » openhab2 #49](https://buildhive.cloudbees.com/job/openhab/job/openhab2/49/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

peuter commented:
Any idea what this is about?

No, sorry. The package "ch.ethz.vppserver.schema.ippclient" does 
contain  the ObjectFactory.class, but I am not familiar with the cups4j 
internals.

buildhive commented:
[openhab » openhab2 #50](https://buildhive.cloudbees.com/job/openhab/job/openhab2/50/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

kaikreuzer commented:
> No, sorry. The package "ch.ethz.vppserver.schema.ippclient" does
contain  the ObjectFactory.class, but I am not familiar with the cups4j
internals.

I found the problem: The cups lib required javax.xml classes, but they were not importer by the bundle and thus not available on the classpath.

kaikreuzer commented:
Thanks for the update regarding the name, it looks good now!

kaikreuzer commented:
Hm, I just merged it including my changes here: https://github.com/openhab/openhab2/commit/8d6318ce3ab304e882147f02fe178b492bdb3e86
But somehow the reference to this PR got lost, so I close it manually.

kaikreuzer commented:
@peuter: May I ask you to create a short documentation for this binding until Sunday (May 24th)?
I have created a short guide on how to do it: https://github.com/eclipse/smarthome/blob/master/docs/sources/development/extensions/bindings/docs.md
As an example how the result could look like, please see https://github.com/eclipse/smarthome/blob/master/addons/binding/org.eclipse.smarthome.binding.hue/README.md.
Thanks!

peuter commented:
@kaikreuzer: As I will be "offline" later this day until next wednesday I can only provide a basic description without configuration examples, as you can see here:
https://github.com/peuter/openhab2/tree/ipp-readme/addons/binding/org.openhab.binding.ipp
If that´s ok with you I would create a pull request.

kaikreuzer commented:
Sure, I will accept anything you can come up with in this remaining minutes ;-)
